### ISSUE TITLE
Hardlinks not supported on Object Cloud Storage

### ISSUE THREAD

**Issue by nemox** [tus/tusd#359]
**Question**
Same as issue #257 . The problem is that the hard links are not supported on IBM Cloud Object Storage. Is there any possible way around this? Unfortunately I am not able to work around this by implementing nfs.

When starting the multipart upload to IBM Cloud Object Storage I get the following error: 
```
[tusd] 2020/03/11 11:54:51 event="ResponseOutgoing" status="500" method="HEAD" path="4997bc26d2363dd1dba6bce3606d1230" error="Lockfile created, but doesn't exist"
```
**Setup details**
Image
```
FROM tusproject/tusd

USER root

RUN apk add --no-cache tree curl \
		&& mkdir /srv/tusd-data/tmp \
		&& mkdir /srv/tusd-data/data \
		&& chown -R tusd:tusd /srv

COPY pre-create post-finish /srv/tusd-hooks/
RUN chmod +x /srv/tusd-hooks/pre-create && chmod +x /srv/tusd-hooks/post-finish \
		&& chown tusd:tusd /srv/tusd-hooks/pre-create && chown tusd:tusd /srv/tusd-hooks/post-finish

EXPOSE 1080

ENTRYPOINT ["tusd"]
CMD ["-hooks-dir=/srv/tusd-hooks"]

USER tusd
```
Deployment
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    component: "tusd"
    app: hasher
    release: hasher
    chart: hasher-0.1.0
    heritage: Helm
  name: hasher-tusd
spec:
  replicas: 3
  selector:
    matchLabels:
      component: "tusd"
      app: hasher
      release: hasher
  strategy: {}
  template:
    metadata:
      labels:
        component: "tusd"
        app: hasher
        release: hasher
        chart: hasher-0.1.0
        heritage: Helm
    spec:
      securityContext:
        runAsUser: 1000
        runAsGroup: 1000
        fsGroup: 1000
      containers:
        - image: de.icr.io/nelaspace/tusd_repo:1.2.0
          name: hasher-tusd
          args:
          - "-behind-proxy"
          - "-hooks-dir=/srv/tusd-hooks"
          - "-upload-dir=/srv/tusd-data/data"
          - "-port=1080"
          - "-verbose"
          ports:
          - containerPort: 1080
            name: tusd-web
          imagePullPolicy: Always
          resources: {}
          envFrom:
            - configMapRef:
                name: svc-configmap
            - secretRef:
                name: cos-write-access
          volumeMounts:
          - name: tusd-uploads
            mountPath: /srv/tusd-data/data
      volumes:
      - name: tusd-uploads
        persistentVolumeClaim:
          claimName: hasher-tusd-pvc
status: {} 
```

Service
```
apiVersion: v1
kind: Service
metadata:
  name: hasher-tusd-service
  labels:
    component: "tusd"
    app: hasher
    release: hasher
    chart: hasher-0.1.0
    heritage: Helm
spec:
  type: NodePort
  ports:
  - port: 1080
    targetPort: 1080
    nodePort: 31080
    protocol: TCP
    name: hasher-tusd
  selector:
    component: "tusd"
    app: hasher
    release: hasher
```
PersistentVolumeClaim
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: hasher-tusd-pvc
  annotations:
    
    ibm.io/auto-create-bucket: "false"
    ibm.io/auto-delete-bucket: "false"
    ibm.io/bucket: nela-uploads
    ibm.io/endpoint: https://s3.private.eu-de.cloud-object-storage.appdomain.cloud
    ibm.io/secret-name: cos-write-access
spec:
  accessModes:
  
  - "ReadWriteOnce"
  resources:
    requests:
      storage: "13Gi"
  storageClassName: "ibmc-s3fs-standard-regional"
```

**Comment by Acconut** [comment]
I don't think there is much we can do. Just for the record: Are soft links supported?
As an alternative, you might be able to use IBM Object Storage over its S3 API (https://cloud.ibm.com/docs/services/cloud-object-storage?topic=cloud-object-storage-compatibility-api) and use tusd's S3Store.

**Comment by nemox** [comment]
Soft links are supported. 
I tried using the IBM S3 API, bt was unable to start tusd server correctly. I tried starting up the tusd server with ```-s3-endpoint``` flag, but it was unable to start with message ```Missing Region```. I thought that region was specific to AWS. 
I tried like this but was unable to make it work:
```-s3-bucket=nela-uploads -s3-endpoint=https://s3.private.eu-de.cloud-object-storage.appdomain.cloud ```

Any help on how to fix this is greatly appreciated.

**Comment by Acconut** [comment]
Can you try setting any value for the `AWS_REGION` environment variable? Maybe the AWS SDK wants that value to not be empty.

**Comment by nemox** [comment]
It did work now! This is what I did, maybe it would be useful for someone else.
I set the ```AWS_REGION``` to the same region of my object store. Further more I had to configure a new set of credentials with HMAC authentication for the IBM Cloud Object Storage. Then from these credentials I se the values ```AWS_SECRET_ACCESS_KEY``` and ```AWS_ACCESS_KEY_ID```. Then it works like a charm! 

Thank you so much for your help, you are awesome!
